File: SJSegmentedScrollView-ObjC/SJSegmentView.m
//
//  SJSegmentView.m
//  Hike
//
//  Created by Pavan Goyal on 02/05/18.
//  Copyright © 2018 Hike Pvt. Ltd. All rights reserved.
//

#import "SJSegmentView.h"
#import "SJUtil.h"

@interface SJSegmentView ()

@end

@implementation SJSegmentView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _kSegmentViewTagOffset = 100;
        _segmentViewOffsetWidth = 10.0;
        _segments = [[NSMutableArray alloc] init];
        _contentSubViewWidthConstraints = [[NSMutableArray alloc] init];
        _controllers = [[NSArray alloc] init];
        _segmentViewHeight = 53.0;
        self.showsHorizontalScrollIndicator = NO;
        self.showsVerticalScrollIndicator = NO;
        self.bounces = NO;
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didChangeSegmentIndex:) name:@"DidChangeSegmentIndex" object:nil];
    }
    return self;
}

- (void)setSelectedSegmentViewColor:(UIColor *)selectedSegmentViewColor {
    _selectedSegmentViewColor = selectedSegmentViewColor;
    self.selectedSegmentView.backgroundColor = selectedSegmentViewColor;
}

- (void)setTitleColor:(UIColor *)titleColor {
    _titleColor = titleColor;
    for (SJSegmentTab *segment in self.segments) {
        [segment titleColor:titleColor];
    }
}

- (void)setSegmentBackgroundColor:(UIColor *)segmentBackgroundColor {
    _segmentBackgroundColor = segmentBackgroundColor;
    for (SJSegmentTab *segment in self.segments) {
        [segment setBackgroundColor:segmentBackgroundColor];
    }
}

- (void)setContentView:(SJContentView *)contentView {
    _contentView = contentView;
    [contentView addObserver:self forKeyPath:@"contentOffset" options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionOld context:nil];
}

- (void)didChangeSegmentIndex:(NSNotification *)notification {
    for (SJSegmentTab *button in self.segments) {
        button.isSelected = NO;
    }
    NSNumber *index = notification.object;
    if (index.integerValue < self.segments.count) {
        SJSegmentTab *button = self.segments[index.integerValue];
        button.isSelected = YES;
    }
}

- (void)dealloc {
    [self.contentView removeObserver:self forKeyPath:@"contentOffset" context:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:@"DidChangeSegmentIndex" object:nil];
}

- (void)setSegmentsView:(CGRect)frame {
    CGFloat segmentWidth = [self widthForSegment:frame];
    [self createSegmentContentView:segmentWidth];
    NSInteger index = 0;
    for (UIViewController *controller in self.controllers) {
        [self createSegmentForController:controller width:segmentWidth index:index];
        index += 1;
    }
    [self createSelectedSegmentView:segmentWidth];
    SJSegmentTab *button = self.segments.firstObject;
    button.isSelected = YES;
}

- (void)createSegmentContentView:(CGFloat)titleWidth {
    self.segmentContentView = [[UIView alloc] initWithFrame:CGRectZero];
    self.segmentContentView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.segmentContentView];
    CGFloat contentViewWidth = titleWidth * self.controllers.count;
    NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[contentView]|" options:0 metrics:nil views:@{@"contentView": self.segmentContentView, @"mainView": self}];
    [self addConstraints:horizontalConstraints];
    self.contentViewWidthConstraint = [NSLayoutConstraint constraintWithItem:self.segmentContentView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:contentViewWidth];
    [self addConstraint:self.contentViewWidthConstraint];
    NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[contentView(==mainView)]|" options:0 metrics:nil views:@{@"contentView": self.segmentContentView, @"mainView": self}];
    [self addConstraints:verticalConstraints];
}

- (void)createSegmentForController:(UIViewController *)controller width:(CGFloat)width index:(NSInteger)index {
    SJSegmentTab *segmentTab = [self getSegmentTabForController:controller];
    segmentTab.tag = (index + self.kSegmentViewTagOffset);
    segmentTab.translatesAutoresizingMaskIntoConstraints = NO;
    [self.segmentContentView addSubview:segmentTab];
    if (self.segments.count == 0) {
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[view]" options:0 metrics:nil views:@{@"view": segmentTab}];
        [self.segmentContentView addConstraints:horizontalConstraints];
    } else {
        SJSegmentTab *previousView = self.segments.lastObject;
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:[previousView]-0-[view]" options:0 metrics:nil views:@{@"view": segmentTab, @"previousView": previousView}];
        [self.segmentContentView addConstraints:horizontalConstraints];
    }
    NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:segmentTab attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:width];
    [self.segmentContentView addConstraint:widthConstraint];
    [self.contentSubViewWidthConstraints addObject:widthConstraint];
    NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[view]-1-|" options:0 metrics:nil views:@{@"view": segmentTab}];
    [self.segmentContentView addConstraints:verticalConstraints];
    [self.segments addObject:segmentTab];
}

- (void)createSelectedSegmentView:(CGFloat)width {
    UIView *segmentView = [[UIView alloc] init];
    segmentView.backgroundColor = self.selectedSegmentViewColor;
    segmentView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.segmentContentView addSubview:segmentView];
    self.selectedSegmentView = segmentView;
    self.xPosConstraints = [NSLayoutConstraint constraintWithItem:segmentView attribute:NSLayoutAttributeLeading relatedBy:NSLayoutRelationEqual toItem:self.segmentContentView attribute:NSLayoutAttributeLeading multiplier:1.0 constant:0.0];
    [self.segmentContentView addConstraint:self.xPosConstraints];
    SJSegmentTab *segment = self.segments.firstObject;
    NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:[view(==segment)]" options:0 metrics:nil views:@{@"view": segmentView, @"segment": segment}];
    [self.segmentContentView addConstraints:horizontalConstraints];
    NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[view(height)]|" options:0 metrics:@{@"height": @(self.selectedSegmentViewHeight)} views:@{@"view": segmentView}];
    [self.segmentContentView addConstraints:verticalConstraints];
}

- (SJSegmentTab *)getSegmentTabForController:(UIViewController *)controller {
    SJSegmentTab *segmentTab = nil;
    if (controller.navigationItem.titleView != nil) {
        segmentTab = [[SJSegmentTab alloc] initWithView:controller.navigationItem.titleView];
    } else {
        if (controller.title) {
            segmentTab = [[SJSegmentTab alloc] initWithTitle:controller.title];
        } else {
            segmentTab = [[SJSegmentTab alloc] initWithTitle:@""];
        }
        segmentTab.backgroundColor = self.segmentBackgroundColor;
        [segmentTab titleColor:self.titleColor];
        [segmentTab titleFont:self.font];
    }
    segmentTab.didSelectSegmentAtIndex = self.didSelectSegmentAtIndex;
    segmentTab.translatesAutoresizingMaskIntoConstraints = NO;
    [segmentTab.heightAnchor constraintEqualToConstant:(self.segmentViewHeight-1)].active = YES;
    return segmentTab;
}

- (CGFloat)widthForSegment:(CGRect)frame {
    CGFloat maxWidth = 0;
    for (UIViewController *controller in self.controllers) {
        CGFloat width = 0.0;
        UIView *view = controller.navigationItem.titleView;
        NSString *title = controller.title;
        if (view) {
            width = view.bounds.size.width;
        } else if (title) {
            width = [SJUtil widthForString:title withConstrainedWidth:CGFLOAT_MAX font:self.font];
        }
        if (width > maxWidth) {
            maxWidth = width;
        }
    }
    CGFloat width = maxWidth + self.segmentViewOffsetWidth;
    CGFloat totalWidth = width * self.controllers.count;
    if (totalWidth < frame.size.width)  {
        maxWidth = frame.size.width / (CGFloat)self.controllers.count;
    } else {
        maxWidth = (CGFloat)width;
    }
    return maxWidth;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if ([change isKindOfClass:[NSDictionary class]]) {
        id old = change[NSKeyValueChangeOldKey];
        id new = change[NSKeyValueChangeNewKey];
        if (old && new) {
            if (![old isEqual:new]) {
                if ([object isKindOfClass:[UIScrollView class]]) {
                    UIScrollView *scrollView = object;
                    CGFloat changeOffset = scrollView.contentSize.width / self.contentSize.width;
                    CGFloat value = scrollView.contentOffset.x / changeOffset;
                    if (!isnan(value)) {
                        CGRect frame = self.selectedSegmentView.frame;
                        frame.origin.x = scrollView.contentOffset.x / changeOffset;
                        self.selectedSegmentView.frame = frame;
                    }
                    CGFloat segmentScrollWidth = self.contentSize.width - self.bounds.size.width;
                    CGFloat contentScrollWidth = scrollView.contentSize.width - scrollView.bounds.size.width;
                    changeOffset = segmentScrollWidth / contentScrollWidth;
                    CGFloat x = scrollView.contentOffset.x * changeOffset;
                    CGFloat y = self.contentOffset.y;
                    self.contentOffset = CGPointMake(x, y);
                }
            }
        }
    }
}

- (void)didChangeParentViewFrame:(CGRect)frame {
    CGFloat segmentWidth = [self widthForSegment:frame];
    CGFloat contentViewWidth = segmentWidth * self.controllers.count;
    self.contentViewWidthConstraint.constant = contentViewWidth;
    for (NSLayoutConstraint *constraint in self.contentSubViewWidthConstraints) {
        constraint.constant = segmentWidth;
    }
    CGFloat changeOffset = self.contentView.contentSize.width / self.contentSize.width;
    CGFloat value = self.contentView.contentOffset.x / changeOffset;
    if (!isnan(value)) {
        self.xPosConstraints.constant = self.selectedSegmentView.frame.origin.x;
        [self layoutIfNeeded];
    }
}

@end

File: SJSegmentedScrollView/Classes/SJSegmentView.swift
//
//  SJSegmentView.swift
//  SJSegmentedScrollView
//
//  Created by Subins Jose on 10/06/16.
//  Copyright © 2016 Subins Jose. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
//    associated documentation files (the "Software"), to deal in the Software without restriction,
//    including without limitation the rights to use, copy, modify, merge, publish, distribute,
//    sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
//    furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all copies or
//    substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
//  LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
//  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import UIKit

class SJSegmentView: UIScrollView {
    
    var selectedSegmentViewColor: UIColor? {
        didSet {
            selectedSegmentView?.backgroundColor = selectedSegmentViewColor
        }
    }
    
    var titleColor: UIColor? {
        didSet {
            for segment in segments {
                segment.titleColor(titleColor!)
            }
        }
    }
    
    var selectedTitleColor: UIColor? {
        didSet {
            for segment in segments {
                segment.selectedTitleColor(selectedTitleColor)
            }
        }
    }
    
    var segmentBackgroundColor: UIColor? {
        didSet {
            for segment in segments {
                segment.backgroundColor = segmentBackgroundColor
            }
        }
    }
    
    var shadow: SJShadow? {
        didSet {
            if let shadow = shadow {
                layer.shadowOffset = shadow.offset
                layer.shadowColor = shadow.color.cgColor
                layer.shadowRadius = shadow.radius
                layer.shadowOpacity = shadow.opacity
                layer.masksToBounds = false;
            }
        }
    }
    
    var font: UIFont?
    var selectedSegmentViewHeight: CGFloat?
    let kSegmentViewTagOffset = 100
    var segmentViewOffsetWidth: CGFloat = 10.0
    var segments = [SJSegmentTab]()
    var segmentContentView: UIView?
    var didSelectSegmentAtIndex: DidSelectSegmentAtIndex?
    var selectedSegmentView: UIView?
    var xPosConstraints: NSLayoutConstraint?
    var contentViewWidthConstraint: NSLayoutConstraint?
    var selectedSegmentViewWidthConstraint: NSLayoutConstraint?
    var contentSubViewWidthConstraints = [NSLayoutConstraint]()
	var controllers: [UIViewController]?
    
    var contentView: SJContentView? {
        didSet {
            contentView!.addObserver(self,
                                     forKeyPath: "contentOffset",
                                     options: [NSKeyValueObservingOptions.new, NSKeyValueObservingOptions.old],
                                     context: nil)
        }
    }
    
    required override init(frame: CGRect) {
        super.init(frame: frame)

		showsHorizontalScrollIndicator = false
		showsVerticalScrollIndicator = false
		bounces = false


		NotificationCenter.default.addObserver(self,
		                                       selector: #selector(SJSegmentView.didChangeSegmentIndex(_:)),
		                                       name: NSNotification.Name("DidChangeSegmentIndex"),
		                                       object: nil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        contentView!.removeObserver(self,
                                    forKeyPath: "contentOffset",
                                    context: nil)
        
        NotificationCenter.default.removeObserver(self,
                                                            name:NSNotification.Name("DidChangeSegmentIndex"),
                                                            object: nil)
    }
    
    @objc func didChangeSegmentIndex(_ notification: Notification) {
        
        //deselect previous buttons
        for button in segments {
            button.isSelected = false
        }
        
        // select current button
        let index = notification.object as? Int

		if index! < segments.count {
			let button = segments[index!]
			button.isSelected = true
		}
    }

    func setSegmentsView(_ frame: CGRect) {

        let segmentWidth = widthForSegment(frame)
        createSegmentContentView(segmentWidth)
        
        var index = 0
        for controller in controllers! {
            
            createSegmentFor(controller, width: segmentWidth, index: index)
            index += 1
        }
        
        createSelectedSegmentView(segmentWidth)
        
        //Set first button as selected
        let button = segments.first!
        button.isSelected = true
    }
    
    func createSegmentContentView(_ titleWidth: CGFloat) {
        
        segmentContentView = UIView(frame: CGRect.zero)
        segmentContentView!.translatesAutoresizingMaskIntoConstraints = false
        addSubview(segmentContentView!)
        
        let contentViewWidth = titleWidth * CGFloat((controllers?.count)!)
        let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|[contentView]|",
                                                                                   options: [],
                                                                                   metrics: nil,
                                                                                   views: ["contentView": segmentContentView!,
                                                                                    "mainView": self])
        addConstraints(horizontalConstraints)
        
        contentViewWidthConstraint = NSLayoutConstraint(item: segmentContentView!,
                                                        attribute: .width,
                                                        relatedBy: .equal,
                                                        toItem: nil,
                                                        attribute: .notAnAttribute,
                                                        multiplier: 1.0,
                                                        constant: contentViewWidth)
        addConstraint(contentViewWidthConstraint!)
        
        
        let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|[contentView(==mainView)]|",
                                                                                 options: [],
                                                                                 metrics: nil,
                                                                                 views: ["contentView": segmentContentView!,
                                                                                    "mainView": self])
        addConstraints(verticalConstraints)
    }
    
    func createSegmentFor(_ controller: UIViewController, width: CGFloat, index: Int) {
        
        let segmentView = getSegmentTabForController(controller)
        segmentView.tag = (index + kSegmentViewTagOffset)
        segmentView.translatesAutoresizingMaskIntoConstraints = false
        segmentContentView!.addSubview(segmentView)
        
        if segments.count == 0 {
            
            let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|[view]",
                                                                                       options: [],
                                                                                       metrics: nil,
                                                                                       views: ["view": segmentView])
            segmentContentView!.addConstraints(horizontalConstraints)
            
        } else {
            
            let previousView = segments.last
            let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:[previousView]-0-[view]",
                                                                                       options: [],
                                                                                       metrics: nil,
                                                                                       views: ["view": segmentView,
                                                                                        "previousView": previousView!])
            segmentContentView!.addConstraints(horizontalConstraints)
        }
        
        let widthConstraint = NSLayoutConstraint(item: segmentView,
                                                 attribute: .width,
                                                 relatedBy: .equal,
                                                 toItem: nil,
                                                 attribute: .notAnAttribute,
                                                 multiplier: 1.0,
                                                 constant: width)
        segmentContentView!.addConstraint(widthConstraint)
        contentSubViewWidthConstraints.append(widthConstraint)
        
        let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|[view]|",
                                                                                 options: [],
                                                                                 metrics: nil,
                                                                                 views: ["view": segmentView])
        segmentContentView!.addConstraints(verticalConstraints)
        segments.append(segmentView)
    }
    
    func createSelectedSegmentView(_ width: CGFloat) {
        
        let segmentView = UIView()
        segmentView.backgroundColor = selectedSegmentViewColor
        segmentView.translatesAutoresizingMaskIntoConstraints = false
        segmentContentView!.addSubview(segmentView)
        selectedSegmentView = segmentView
        
        xPosConstraints = NSLayoutConstraint(item: segmentView,
                                             attribute: .leading,
                                             relatedBy: .equal,
                                             toItem: segmentContentView!,
                                             attribute: .leading,
                                             multiplier: 1.0,
                                             constant: 0.0)
        segmentContentView!.addConstraint(xPosConstraints!)
        
        let segment = segments.first
        let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:[view(==segment)]",
                                                                                   options: [],
                                                                                   metrics: nil,
                                                                                   views: ["view": segmentView,
                                                                                    "segment": segment!])
        segmentContentView!.addConstraints(horizontalConstraints)
        
        let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:[view(height)]|",
                                                                                 options: [],
                                                                                 metrics: ["height": selectedSegmentViewHeight!],
                                                                                 views: ["view": segmentView])
        segmentContentView!.addConstraints(verticalConstraints)
    }
    
    func getSegmentTabForController(_ controller: UIViewController) -> SJSegmentTab {

		var segmentTab: SJSegmentTab?

		if controller.navigationItem.titleView != nil {
			segmentTab = SJSegmentTab.init(view: controller.navigationItem.titleView!)
		} else {

			if let title = controller.title {
				segmentTab = SJSegmentTab.init(title: title)
			} else {
				segmentTab = SJSegmentTab.init(title: "")
			}

			segmentTab?.backgroundColor = segmentBackgroundColor
			segmentTab?.titleColor(titleColor!)
            segmentTab?.selectedTitleColor(selectedTitleColor!)
			segmentTab?.titleFont(font!)
		}

		segmentTab?.didSelectSegmentAtIndex = didSelectSegmentAtIndex

        return segmentTab!
    }

	func widthForSegment(_ frame: CGRect) -> CGFloat {

		var maxWidth: CGFloat = 0
		for controller in controllers! {

			var width: CGFloat = 0.0
			if let view = controller.navigationItem.titleView {
				width = view.bounds.width
			} else if let title = controller.title {

				width = title.widthWithConstrainedWidth(.greatestFiniteMagnitude,
				                                        font: font!)
			}

			if width > maxWidth {
				maxWidth = width
			}
		}

		let width = Int(maxWidth + segmentViewOffsetWidth)
		let totalWidth = width * (controllers?.count)!
		if totalWidth < Int(frame.size.width)  {
			maxWidth = frame.size.width /  CGFloat((controllers?.count)!)
		} else {
			maxWidth = CGFloat(width)
		}

		return maxWidth
	}
    
	override func observeValue(forKeyPath keyPath: String?,
	                           of object: Any?,
	                           change: [NSKeyValueChangeKey : Any]?,
	                           context: UnsafeMutableRawPointer?) {
        
        if let change = change as [NSKeyValueChangeKey : AnyObject]? {
            if let old = change[NSKeyValueChangeKey.oldKey], let new = change[NSKeyValueChangeKey.newKey] {
                if !(old.isEqual(new)) {
                    //update selected segment view x position
                    let scrollView = object as? UIScrollView
                    var changeOffset = (scrollView?.contentSize.width)! / contentSize.width
                    let value = (scrollView?.contentOffset.x)! / changeOffset
                    
                    if !value.isNaN {
                        selectedSegmentView?.frame.origin.x = (scrollView?.contentOffset.x)! / changeOffset
                    }
                    
                    //update segment offset x position
                    let segmentScrollWidth = contentSize.width - bounds.width
                    let contentScrollWidth = scrollView!.contentSize.width - scrollView!.bounds.width
                    changeOffset = segmentScrollWidth / contentScrollWidth
                    contentOffset.x = (scrollView?.contentOffset.x)! * changeOffset
                }
            }
        }
    }
    
    func didChangeParentViewFrame(_ frame: CGRect) {
        
        let segmentWidth = widthForSegment(frame)
        let contentViewWidth = segmentWidth * CGFloat((controllers?.count)!)
        contentViewWidthConstraint?.constant = contentViewWidth
        
        for constraint in contentSubViewWidthConstraints {
            constraint.constant = segmentWidth
        }
        
        let changeOffset = (contentView?.contentSize.width)! / contentSize.width
        let value = (contentView?.contentOffset.x)! / changeOffset
        
        if !value.isNaN {
            if UIView.userInterfaceLayoutDirection(for: self.semanticContentAttribute) == .rightToLeft {
                let newIndex = contentView?.pageIndex ?? 0
                xPosConstraints!.constant = CGFloat(newIndex) * segmentWidth
            } else {
                xPosConstraints!.constant = (selectedSegmentView?.frame.origin.x)!
            }
            
            layoutIfNeeded()
        }
    }
}
